rslib/src/cloze.rs

```diff
@@ -31,99 +19,262 @@ lazy_static! {
     pub const OPENING_TAG: usize = 1;
     pub const INNER_TEXT: usize = 2;
     pub const CLOSING_TAG: usize = 3;
 }
 
-pub fn reveal_cloze_text(text: &str, cloze_ord: u16, question: bool) -> Cow<str> {
-    let mut cloze_ord_was_in_text = false;
-
-    let output = CLOZE.replace_all(text, |caps: &Captures| {
-        let captured_ord = caps
-            .get(cloze_caps::ORD)
-            .unwrap()
-            .as_str()
-            .parse()
-            .unwrap_or(0);
-
-        let text = caps.get(cloze_caps::TEXT).unwrap().as_str().to_owned();
-        if captured_ord != cloze_ord {
-            // other cloze deletions are unchanged
-            return text;
-        } else {
-            cloze_ord_was_in_text = true;
+/// States for cloze parsing state machine
+#[derive(PartialEq, Copy, Clone)]
+enum State {
+    Root,
+    Text,
+    Open,
+    Open2,
+    COpen,
+    Ord,
+    TextOpen1,
+    TextOpen2,
+    TextClose1,
+    HintOpen1,
+    HintOpen2,
+    Hint,
+    HintClose1,
+    Close,
+    Abandon,
+}
+/// Struct for storing data for one cloze
+struct Cloze {
+    buf: String,
+    ord_str: String,
+    text: String,
+    hint: String,
+}
+
+impl Cloze {
+    /// Create empty Cloze struct
+    fn new() -> Self {
+        Self {
+            buf: String::new(),
+            ord_str: String::new(),
+            text: String::new(),
+            hint: String::new(),
+        }
+    }
+    /// Create Cloze struct with buf set to c
+    fn new_from(c: char) -> Self {
+        Self {
+            buf: String::from(c),
+            ord_str: String::new(),
+            text: String::new(),
+            hint: String::new(),
         }
+    }
+}
+
+/// Process char for next state
+fn process_char(state: State, c: char) -> State {
+    use self::State::*;
+    match (state, c) {
+        (Root, '{') => Open,
+        (Open, '{') => Open2,
+        (Open, _) => Abandon,
+        (Open2, '{') => Open2,
+        (Open2, 'c') => COpen,
+        (Open2, _) => Abandon,
+        (COpen, '0'..='9') => Ord,
+        (Ord, '0'..='9') => Ord,
+        (Ord, ':') => TextOpen1,
+        (Ord, _) => Abandon,
+        (TextOpen1, ':') => TextOpen2,
+        (TextOpen1, _) => Abandon,
+        (TextOpen2, ':') => HintOpen1,
+        (TextOpen2, '}') => TextClose1,
+        (TextOpen2, _) => Text,
+        (Text, '{') => Open,
+        (Text, ':') => HintOpen1,
+        (Text, '}') => TextClose1,
+        (Text, _) => Text,
+        (HintOpen1, ':') => HintOpen2,
+        (HintOpen1, '}') => TextClose1,
+        (HintOpen1, _) => Text,
+        (HintOpen2, '}') => HintClose1,
+        (HintOpen2, _) => Hint,
+        (Hint, '}') => HintClose1,
+        (Hint, _) => Hint,
+        (TextClose1 | HintClose1, '}') => Close,
+        (TextClose1, _) => Text,
+        (HintClose1, _) => Hint,
+        _ => Root,
+    }
+}
+
+/// Minimal encoding of string for storage in attribute (", &, \n, <, >)
+pub fn encode_attribute(text: &str) -> Cow<str> {
```

Comment:
I think we should be using htmlescape::encode_attribute() here. See the review on https://github.com/ankitects/anki/pull/1968

Reply:
Ok